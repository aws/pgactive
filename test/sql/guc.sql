-- Allow commands via ALTER SYSTEM SET, config file, ALTER DATABASE set, etc

ALTER SYSTEM
  SET bdr.skip_ddl_replication = on;

-- The check for per-database settings only occurs when you're on that
-- database, so we don't block the setting on another DB and the user
-- has to undo it later.
SELECT current_database();

-- Should be ok
ALTER DATABASE postgres
  SET bdr.skip_ddl_replication = on;

-- Should fail
ALTER DATABASE postgres
  SET bdr.skip_ddl_replication = off;

-- An ERROR setting a GUC doesn't stop the connection to the DB
-- from succeeding though.
\c postgres
SELECT current_database();

ALTER DATABASE postgres
  RESET bdr.skip_ddl_replication;

\c postgres
SELECT current_database();

\c regression
SELECT current_database();

-- This is true even when you ALTER the current database, so this
-- commits fine, but switching back to the DB breaks:
ALTER DATABASE regression
  SET bdr.skip_ddl_replication = off;

\c postgres
SELECT current_database();
-- so this will report an error, but we'll still successfully connect to the DB.
\c regression
SELECT current_database();

-- and fix the GUC
ALTER DATABASE regression
  RESET bdr.skip_ddl_replication;

\c regression
SELECT current_database();

-- Fixed.



-- Explicit "off" is Not OK
ALTER DATABASE regression
  SET bdr.skip_ddl_replication = off;

-- Unless at the system level
ALTER SYSTEM
  SET bdr.skip_ddl_replication = off;

ALTER SYSTEM
  RESET bdr.skip_ddl_replication;

-- Per-user is OK

ALTER USER super
  SET bdr.skip_ddl_replication = on;

-- Unless not permitted
ALTER USER super
  SET bdr.skip_ddl_replication = off;

ALTER USER super
  RESET bdr.skip_ddl_replication;

-- Per session is OK
SET bdr.skip_ddl_replication = on;

-- Unless values are not permitted
SET bdr.skip_ddl_replication = off;

RESET bdr.skip_ddl_replication;;
