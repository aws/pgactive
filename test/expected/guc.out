-- Allow commands via ALTER SYSTEM SET, config file, ALTER DATABASE set, etc
ALTER SYSTEM
  SET pgactive.skip_ddl_replication = on;
-- The check for per-database settings only occurs when you're on that
-- database, so we don't block the setting on another DB and the user
-- has to undo it later.
SELECT current_database();
 current_database 
------------------
 regression
(1 row)

-- Should be ok
ALTER DATABASE postgres
  SET pgactive.skip_ddl_replication = on;
-- Should fail
ALTER DATABASE postgres
  SET pgactive.skip_ddl_replication = off;
WARNING:  unsafe pgactive configuration options can not be disabled locally
DETAIL:  The pgactive option pgactive.skip_ddl_replication should only be disabled globally.
HINT:  See the manual for information on these options. Using them without care can break replication.
ERROR:  invalid value for parameter "pgactive.skip_ddl_replication": 0
-- An ERROR setting a GUC doesn't stop the connection to the DB
-- from succeeding though.
\c postgres
SELECT current_database();
 current_database 
------------------
 postgres
(1 row)

ALTER DATABASE postgres
  RESET pgactive.skip_ddl_replication;
\c postgres
SELECT current_database();
 current_database 
------------------
 postgres
(1 row)

\c regression
SELECT current_database();
 current_database 
------------------
 regression
(1 row)

-- This is true even when you ALTER the current database, so this
-- commits fine, but switching back to the DB breaks:
ALTER DATABASE regression
  SET pgactive.skip_ddl_replication = off;
WARNING:  unsafe pgactive configuration options can not be disabled locally
DETAIL:  The pgactive option pgactive.skip_ddl_replication should only be disabled globally.
HINT:  See the manual for information on these options. Using them without care can break replication.
ERROR:  invalid value for parameter "pgactive.skip_ddl_replication": 0
\c postgres
SELECT current_database();
 current_database 
------------------
 postgres
(1 row)

-- so this will report an error, but we'll still successfully connect to the DB.
\c regression
SELECT current_database();
 current_database 
------------------
 regression
(1 row)

-- and fix the GUC
ALTER DATABASE regression
  RESET pgactive.skip_ddl_replication;
\c regression
SELECT current_database();
 current_database 
------------------
 regression
(1 row)

-- Fixed.
-- Explicit "off" is Not OK
ALTER DATABASE regression
  SET pgactive.skip_ddl_replication = off;
WARNING:  unsafe pgactive configuration options can not be disabled locally
DETAIL:  The pgactive option pgactive.skip_ddl_replication should only be disabled globally.
HINT:  See the manual for information on these options. Using them without care can break replication.
ERROR:  invalid value for parameter "pgactive.skip_ddl_replication": 0
-- Unless at the system level
ALTER SYSTEM
  SET pgactive.skip_ddl_replication = off;
ALTER SYSTEM
  RESET pgactive.skip_ddl_replication;
-- Per-user is OK
ALTER USER super
  SET pgactive.skip_ddl_replication = on;
-- Unless not permitted
ALTER USER super
  SET pgactive.skip_ddl_replication = off;
WARNING:  unsafe pgactive configuration options can not be disabled locally
DETAIL:  The pgactive option pgactive.skip_ddl_replication should only be disabled globally.
HINT:  See the manual for information on these options. Using them without care can break replication.
ERROR:  invalid value for parameter "pgactive.skip_ddl_replication": 0
ALTER USER super
  RESET pgactive.skip_ddl_replication;
-- Per session is OK
SET pgactive.skip_ddl_replication = on;
-- Unless values are not permitted
SET pgactive.skip_ddl_replication = off;
WARNING:  unsafe pgactive configuration options can not be disabled locally
DETAIL:  The pgactive option pgactive.skip_ddl_replication should only be disabled globally.
HINT:  See the manual for information on these options. Using them without care can break replication.
ERROR:  invalid value for parameter "pgactive.skip_ddl_replication": 0
RESET pgactive.skip_ddl_replication;;
